File: problem/Sort_Matrix_Diagonals/sortMatrix.c
int cmpAsc(const void *a, const void *b) { return (*(int *)a - *(int *)b); }
int cmpDesc(const void *a, const void *b) { return (*(int *)b - *(int *)a); }

int **sortMatrix(int **grid, int gridSize, int *gridColSize, int *returnSize, int **returnColumnSizes)
{
    *returnSize = gridSize;
    *returnColumnSizes = gridColSize;
    int n = gridSize, m = gridColSize[0];
    int diagCount = n + m - 1;
    int *sizes = (int *)calloc(diagCount, sizeof(int));
    int **diags = (int **)malloc(diagCount * sizeof(int *));

    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            sizes[i - j + m - 1]++;

    for (int k = 0; k < diagCount; k++)
    {
        diags[k] = (int *)malloc(sizes[k] * sizeof(int));
        sizes[k] = 0;
    }
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            diags[i - j + m - 1][sizes[i - j + m - 1]++] = grid[i][j];

    for (int k = 0; k < diagCount; k++)
    {
        if (k < m - 1)
            qsort(diags[k], sizes[k], sizeof(int), cmpAsc);
        else
            qsort(diags[k], sizes[k], sizeof(int), cmpDesc);
        sizes[k] = 0;
    }
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            grid[i][j] = diags[i - j + m - 1][sizes[i - j + m - 1]++];
    return grid;
}

File: problem/Sort_Matrix_Diagonals/sortMatrix.js
/**
 * @param {number[][]} grid
 * @return {number[][]}
 */
var sortMatrix = function(grid) {
    const n = grid.length, m = grid[0].length;
    const diags = {};

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            const key = i - j;
            if (!(key in diags)) diags[key] = [];
            diags[key].push(grid[i][j]);
        }
    }

    for (const key in diags) {
        if (key < 0) diags[key].sort((a,b)=>a-b);
        else diags[key].sort((a,b)=>b-a);
    }

    const idx = {};
    for (const key in diags) idx[key] = 0;

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            const key = i - j;
            grid[i][j] = diags[key][idx[key]++];
        }
    }
    return grid;
};